### ISSUE TITLE
multi process; scale 

### ISSUE THREAD

**Issue by weihaosky** [aosokin/os2d#11]
Hi, nice work! I have two questions about the code:

1. Could the training be run with multiple GPUs and the dataloader be run with multiple workers? That would be much faster
2. How should I set the dataset scale? What does it mean? I saw that the scale "defining the expected object size" but still cannot understand what it is and how should I set it. Should I change it when I apply your trained model to some new images?

Many thanks!

**Comment by aosokin** [comment]
Hi, thanks for your interest!
1.  Multi-gpu training should be possible to add with torch.nn.parallel.DistributedDataParallel but we did not implement this functionality.
2. To set the scale parameter, you need to know the size of the typical objects in your data (by the size of the box I mean the square root of its area). Then, you need to resize your images in such a way that the typical scale matches the scale used for training (large deviations from 240 that we used can require architecture changes as well). Please, see the script https://github.com/aosokin/os2d/blob/master/data/get_dataset_scales.py for the scale estimation procedure.

Best,
Anton

**Comment by weihaosky** [comment]
Thanks for the reply!
Another question is that the warning "Number of positives in a batch cannot be zero, set to 1!" keeps popping up. Is this normal? Why is this? What impact would it have, e.g. the performance?

**Comment by aosokin** [comment]
> Another question is that the warning "Number of positives in a batch cannot be zero, set to 1!" keeps popping up. Is this normal? Why is this? What impact would it have, e.g. the performance?

This message indicates that in your batch there are no positives. The number of positives is directly used in the normalization of ContrastiveLoss. RLL loss does not depend on this.

However, if you see this message often it means that you often have batches without positives (due to subsamling or some other reason), which I think is a suspicious sign (but not necessarily bad). You might want to try changing batch sampling parameters (e.g. use a larger patch or more classes in a batch) or use hard patch mining.

**Comment by weihaosky** [comment]
Thanks.
It seems the no-positive is worsened by the “remap_classification_targets”. What does this do? What are its advantages

**Comment by aosokin** [comment]
> It seems the no-positive is worsened by the “remap_classification_targets”. What does this do? What are its advantages

At the core of our method, we have a network that produces a transformation aiming to align the class image to the specific location of the input image (defined by an anchor). This transformation can significantly change the detection box so the original positive-negative assignment can go wrong. The flag remap_classification_targets makes the method to recompute positive-negative assignment after the transformation is applied. Btw, there is a different set of IoU thresholds that is with remap_classification_targets, and those are by default harsher to define positives.

**Comment by Usama3059** [comment]
Hi  @aosokin  Its really amazing work you have done the only question is I have been using this for one of my task but it is creating two or three box on the same object your suggestions on this I felt maybe more checking with nms removes this.
kindly let me know.

**Comment by aosokin** [comment]
Hi @Usama3059 , intersecting detections should be taken care of by NMS. Please, note that we have a flag on whether to do NMS across classes: eval.nms_across_classes. By default, it is False, so you might see detections of different categories at the same location. Otherwise, you can try tuning eval.nms_iou_threshold and eval.nms_score_threshold to get rid of some detections.
Best,
Anton

**Comment by WeitaoVan** [comment]
@aosokin Hi aosokin. I have a similar issue. When I use ResNet50 backbone, everything is okay. But when I switched to HRNet, most training iterations show "loss = 0.000".  (all other settings are the same as ResNet50 backbone)
I have correctly set the corresponding size receptive field and stride. I suspect this has something to do with the “remap_classification_targets”.

I would like to hear your opinions on this matter. Thanks！

**Comment by aosokin** [comment]
@WeitaoVan Hi Weitao, it might be the case that when you switch backbone the correlation maps start looking very different and then the transformation network goes mad and stops working.
My debug strategy would start with checking that correlation maps computed at
https://github.com/aosokin/os2d/blob/96c488b3979c505b92245ccdab941fb01fdf99de/os2d/modeling/head.py#L342
look reasonable on the pretrained backbone (you can compare them to the ones from ResNet-50): they should be smooth and should show peaks at locations corresponding to visually similar locations. 
The next step would be to see what the pretrained transformation network does: as in our V2 model it should be working reasonably without any fine-tuning.

Best,
Anton

**Comment by aosokin** [comment]
@WeitaoVan another idea: if for some reason the transformation network does not work with your backbone you can potentially retrain it from scratch for your backbone directly using the repos of Ignacio Rocco et al.: https://github.com/ignacio-rocco/cnngeometric_pytorch https://github.com/ignacio-rocco/weakalign

**Comment by WeitaoVan** [comment]
@aosokin your suggestions are much appreciated, I will look into it. However, the strange thing is how can the loss be 0.000? And it happens in about 8 iterations out of 10. 
The only reason I can think of is that by changing the backbone, something is wrong with the label preparation. The classification targets are somehow made all negatives. 
I will verify it by printing the classification targets.
But how can the labels be affected by changing backbones?

**Comment by WeitaoVan** [comment]
I forgot to mention that the training is on INSTRE dataset. So the only loss is the classification loss as loc_weight is 0

**Comment by aosokin** [comment]
> The classification targets are somehow made all negatives.

I think this can happen if the transformation network is totally off (e.g., because of very different correlation maps), which results in very bad detection boxes. After this, the system relabels targets (as the flag ``remap_classification_targets'' is on) and labels everything as negatives. This training situation looks more similar to our V1 (not V2) where the transformation net was trained from scratch. In this situation, one should probably set remap_classification_targets to False and turn on the localization loss.
After this, it might be possible to fine-tune in the V2-like mode.
